### PR TITLE
Add secret mount to operator

### PR DESCRIPTION
**Pull request by Shinzu**
First initial commit for adding  cloud secrets to non cloud environment.

It ties to address #198

### CODE REVIEW

**Comment by redbaron** [comment]
Have you seen https://github.com/zalando/postgres-operator/pull/481 ?

**Comment by Shinzu** [comment]
yes but i think this tackles an other problem

File: docs/administrator.md [hunk 1]
```diff
@@ -346,3 +346,19 @@ To read the latter log entry with the escaped characters rendered, view it
...
 noise because the `PodTemplate` used by the operator is yet to be updated
 with the default values used internally in Kubernetes.
+
+## Access to cloud resources from clusters in non cloud environment
+
+To access cloud resources like S3 from a cluster in a bare metal setup you can use
+`additional_secret_mount` and `additional_secret_mount_path` config parameters.
+With this you can provision cloud credentials to the containers in the pods of the StatefulSet.
+This works this way that it mounts a volume from the given secret in the pod and this can
+then accessed in the container over the configured mount path. Via [Custum Pod Environment Variables](## Custom Pod Environment Variables)
```
Comment (FxKu) [inline docs/administrator.md]: ... of the StatefulSet.
A volume is mounted from the given secret in the pod and can then be accessed in the container over the configured mount path. Via [Custom Pod Environment Variables](## Custom Pod Environment Variables) ... 

btw: link doesn't work for me

File: docs/administrator.md [hunk 2]
```diff
@@ -346,3 +346,19 @@ To read the latter log entry with the escaped characters rendered, view it
...
+you can then point the different cloud sdk's (aws, google etc.) to this mounted secret.
+With this credentials the cloud sdk can then access cloud resources to upload logs etc.
+
+This secrets can be pre provisioned in different ways:
```
Comment (FxKu) [inline docs/administrator.md]: Keep a singular: A secret can be ...

File: docs/reference/operator_parameters.md [hunk 3]
```diff
@@ -366,6 +366,12 @@ either. In the CRD-based configuration those options are grouped under the
 * **aws_region**
   AWS region used to store ESB volumes. The default is `eu-central-1`.
 
+* **additional_secret_mount**
+  Additional Secret (aws or gcp credentials) to mount in the the pod. The default is empty.
```
Comment (FxKu) [inline docs/reference/operator_parameters.md]: in the pod

File: docs/administrator.md (hunk 1 above)
Comment (Shinzu) [reply docs/administrator.md]: fixed

File: docs/reference/operator_parameters.md (hunk 3 above)
Comment (Shinzu) [reply docs/reference/operator_parameters.md]: fixed

File: docs/administrator.md (hunk 2 above)
Comment (Shinzu) [reply docs/administrator.md]: fixed

**Comment by redbaron** [comment]
We keep poorly reinventing pod spec, when is the time to stop? https://github.com/zalando/postgres-operator/issues/479

**Comment by Shinzu** [comment]
@redbaron i'm with you, in the end this is just convenience function for user who don't want to create a whole pod spec

**Comment by erthalion** [comment]
As we discussed in #479, most likely we would need to have both "modes" - full spec and not full spec. In this sense I believe it makes sense to have this feature for the second mode, what do you think?

File: pkg/cluster/k8sres_test.go [hunk 4]
```diff
@@ -129,3 +129,74 @@ func TestShmVolume(t *testing.T) {
...
+			if volumeMountName != additionalSecretMount {
+				t.Errorf("%s %s: Expected mount %s was not created, have %s instead",
+					testName, tt.subTest, additionalSecretMount, volumeMountName)
+			}
+		}
+
+		volsuma := len(tt.podSpec.Containers[0].VolumeMounts)
+
+		if volsuma != volsumb+1 {
+			t.Errorf("Got not expected number of VolumeMounts: got %v instead of %v",
```
Comment (FxKu) [inline pkg/cluster/k8sres_test.go]: "Unexpected number ...

btw: I find `volsumb` a strange name. Maybe `numMounts` for b and `numMountsCheck` for a?

File: manifests/postgresql-operator-default-configuration.yaml [hunk 5]
```diff
@@ -61,6 +61,8 @@ configuration:
     # log_s3_bucket: ""
     # kube_iam_role: ""
     aws_region: eu-central-1
+    # additional_secret_mount: 'some-secret-name'
+    # additional_secret_mount_path: '/some/dir'
```
Comment (FxKu) [inline manifests/postgresql-operator-default-configuration.yaml]: does it really work with single quotes? In the ConfigMap I can see other examples with them but not in the `OperatorConfiguration`. Other the other way around: Would `"` work for ConfigMap, too.? If yes, change it also in the `values.yaml` file for Helm chart. The Helm templates for ConfigMap and CRD can stay as they are.

File: pkg/cluster/k8sres_test.go (hunk 4 above)
Comment (Shinzu) [reply pkg/cluster/k8sres_test.go]: renamed variables

File: manifests/postgresql-operator-default-configuration.yaml (hunk 5 above)
Comment (Shinzu) [reply manifests/postgresql-operator-default-configuration.yaml]: using now double quotes

**Comment by Jan-M** [comment]
Looks good to me. @sdudoladov @FxKu

**Comment by FxKu** [comment]
@Shinzu could you please rebase and resolve the conflicts?

File: pkg/cluster/k8sres.go [hunk 6]
```diff
@@ -993,6 +1001,28 @@ func addShmVolume(podSpec *v1.PodSpec) {
...
+	volumes := append(podSpec.Volumes, v1.Volume{
+		Name: additionalSecretMount,
+		VolumeSource: v1.VolumeSource{
+			Secret: &v1.SecretVolumeSource{
+				SecretName: additionalSecretMount,
+			},
+		},
+	})
+
+	for i := range podSpec.Containers {
```
[the remark this answers is not in the document]
Comment (Shinzu) [reply pkg/cluster/k8sres.go]: yes it is intentional

**Comment by sdudoladov** [comment]
@Shinzu can you please rebase and then we will merge this PR

**Comment by Shinzu** [comment]
rebase done

**Comment by sdudoladov** [comment]
:+1:

**Comment by FxKu** [comment]
hm, for some reason the config parameters are in single quotes again?

**Comment by Shinzu** [comment]
@FxKu @sdudoladov re applied the lost changes

**Comment by FxKu** [comment]
:+1:

**Comment by sdudoladov** [comment]
:+1:

**Comment by KarstenSiemer** [comment]
Hi!
Can somebody tell me how the secret has to look like exactly? There is no use to saying there is a secret with a json file in it you can attach but putting no information on how it has to look like.
The documentation of WAL-G is only point out env variables and spilo is equally silent about how to pass a json.
Thanks!

**Comment by Shinzu** [comment]
hello, this file just looks like a normal credential file for aws or gcp eg https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html

**Comment by KarstenSiemer** [comment]
i still do not completely see the full image for getting wal-e to work with baremetal clusters.
> Via Custom Pod Environment Variables you can point different cloud SDK's (AWS, GCP etc.) to this mounted secret, e.g. to access cloud resources for uploading logs etc.

>AWS, GCP etc.

So one mounts a file with the credentials, this is placed inside the pod but unused, unless i configure env vars to pick up that file.

But, when looking at the documentation of wal-e, only gcp has a variable ("GOOGLE_APPLICATION_CREDENTIALS") for that. AWS doesn't.
How does one pass the credentials file to the aws connector?

**Comment by whereismyjetpack** [comment]
The behavior of this PR is slightly different than the behavior of injecting these values from the configmap. Mainly, the secret object must live in the namespace where the postgresql object is, and not the operator..

Ideally, for our usecase, we'd specify cloud credentials in the postgres-operator namespace, and the operator would add them to the existing secret that gets created, or to a new cloud credentials secret object.

**Pull request merged**